src/datasource/gateway.js

```diff
@@ -0,0 +1,243 @@
+      method: 'post',
+      data: {
+        address,
+        height,
+      }
+    }),
+    getTransaction: (id) => ({
+      url: '/v1/query/tx',
+      method: 'post',
+      data: {
+        hash: id,
+      }
+    }),
+    getApp: (address, height) => ({
+      url: '/v1/query/app',
+      method: 'post',
+      data: {
+        address,
+        height,
+      }
+    }),
+    getNode: (address, height) => ({
+      url: '/v1/query/node',
+      method: 'post',
+      data: {
+        address,
+        height,
+      }
+    }),
+    getAccountTxs: (address, received, prove, page, per_page) => ({
+      url: '/v1/query/accounttxs',
+      method: 'post',
+      data: {
+        address,
+        prove,
+        received,
+        page,
+        per_page,
+      }
+    }),
+    sendRawTx: (fromAddress, tx) => ({
+      url: '/v1/client/rawtx',
+      method: 'post',
+      data: {
+        address: fromAddress,
+        raw_hex_bytes: tx,
+      }
+    })
+  }
+
+  use(provider) {
+    this.provider = provider;
+    return this;
+  }
+
+  // looks ugly with the ifs
+  // but the gateway constantly responds with 200
+  // and responds with errors in response.data
+  // in a non-consistent form.
+  parseSuccessfulResponse = (response) => {
+    
+    if (typeof response.data === 'string' && response.data.indexOf('Method Not Allowed') > -1) {
+      throw new Error('Method Not Allowed')
+    }
+
+    if (response.data.code && response.data.code !== 200) {
+      throw response.data;
+    }
+
+    return response.data;
+  }
+
+  parseErrorResponse = (error) => {
+    if (error.response && error.response.data && error.response.data.error) {
+      throw error.response.data.error;
+    }
+
+    if (typeof error === 'string') {
+      throw new Error(error)
+    }
+
+    throw error;
+
+  }
+
+  perform = async (requestName, ...args) => {
+    const reqConfig = typeof this.requests[requestName] === 'function' ?
+      this.requests[requestName](...args) :
+      this.requests[requestName];
+
+    const response = await this
+      .provider
+      .http
+      .request(reqConfig)
+      .then(this.parseSuccessfulResponse)
+      .catch(this.parseErrorResponse);
+
+    return response;
+  }
+  
+  // does not really need to be bound to `this`, but keeping it for semantics' sake.
+  // arguments explicit forwardong for clear signature lookup, avoid using `...args` 
+  _getBalance = (address, height) => this.perform.call(this, 'getBalance', address, height) 
+  _getTransaction = (id) => this.perform.call(this, 'getTransaction', id)
+  _getApp = (address, height) => this.perform.call(this, 'getApp', address, height)
+  _getNode = (address, height) => this.perform.call(this, 'getNode', address, height)
+  _getAccountTxs = (address, received, prove, page, per_page) => this.perform.call(this, 'getAccountTxs', address, received, prove, page, per_page)
+  _sendRawTx = (fromAddress, tx) => this.perform.call(this, 'sendRawTx', fromAddress, tx)
```

Comment:
I think `_functionName` is confusing, I will switch to:

```
 requestHandlers = {
    handlerName
  }
 ```
 
 If you are wondering why at all, it is to allow for easier interception pre and post request. Check the following lines in `query = `

Reply:
Definitely appreciate this change :)